src/system/system_daemon.c

```diff
@@ -45,13 +49,27 @@ task_fn_t task_fns[4] = {_opcontrol_task, _autonomous_task, _disabled_task, _com
 
 extern void ser_output_flush(void);
 
+void add_daemon_hook(generic_fn_t hook) {
+	// TODO: make thread safe
+	if (unlikely(hook_list == NULL)) {
+		hook_list = linked_list_init();
+	} 
+	linked_list_append_func(hook_list, hook);
+}
+
+static void call_hook(ll_node_s_t* node, void* data) {
+	(void) data;
+	node->payload.func();
+}
+
 // does the basic background operations that need to occur every 2ms
 static inline void do_background_operations() {
 	port_mutex_take_all();
 	ser_output_flush();
 	rtos_suspend_all();
 	vexBackgroundProcessing();
 	rtos_resume_all();
+	linked_list_foreach(hook_list, &call_hook, NULL);
 	vdml_background_processing();
 	port_mutex_give_all();
 }
```

Comment:
I'm not sold on this order - I frankly don't know what vdml_background_processing is, but it sounds like something that should happen first.

Also, any user code should definitely not be called while all port mutexes are locked, because thats obviously going to deadlock, as every device API function requires locking a port mutex.

I think the `linked_list_foreach` should be at the end, after `port_mutex_give_all()`.